# Predicting Poverty W/ Anonymized Data

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
poverty_train =  pd.read_csv('LabTest_train_data.csv')
poverty_test = pd.read_csv('LabTest_test_data.csv')

# Exploratory Data Analysis

In [3]:
pd.set_option('display.max_columns', None)
poverty_train.head()

id wBXbHZmp SlDKnCuu KAJOWiiw DsKacCdL rtPrBBPl tMJrvvut jdetlNNF  \
0  46107    JhtDR    GUusz    TuovO    ZYabk    feupP    PHMVg    NDTCU   
1  82739    JhtDR    GUusz    TuovO    ZYabk    feupP    PHMVg    NDTCU   
2   9646    JhtDR    GUusz    BIZns    ZYabk    uxuSS    PHMVg    NDTCU   
3  16463    JhtDR    alLXR    TuovO    ZYabk    feupP    PHMVg    NDTCU   
4  98438    JhtDR    alLXR    BIZns    ZYabk    feupP    PHMVg    NDTCU   

  maLAYXwi vwpsXRGk sArDRIyX goxNwvnG TYhoEiNm bgfNZfcj sYIButva VZtBaoXL  \
0    cLAGr    XAmOF    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   
1    sehIp    lwCkE    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   
2    sehIp    qNABl    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    usVYt   
3    cLAGr    NdlDR    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   
4    sehIp    bUhyU    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    usVYt   

  GUvFHPNA zFkComtB fxbqfEWb HDMHzGif zzwlWZZC nGTepfos DxLvCGgv CbABToOI  \
0    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    GEOXc    oHRvd   
1    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    ltpcC    VbCFp    tIJga   
2    JGQEQ    JETTQ    WobIQ    PSbhz    Xtgis    nvuhC    GEOXc    tIJga   
3    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    GEOXc    oHRvd   
4    JGQEQ    RnjPp    WobIQ    PSbhz    CtanB    nvuhC    VbCFp    tIJga   

  qgMygRvX fhKiXuMY uSKnVaKV hESBInAl nzTeWUeM BbKZUYsB UCnazcxd hTraVEWP  \
0    Bdpcw    nJuqx    uPkpA    ZZLWA    oKdrT    VxayU    GELir    ZkQxz   
1    Bdpcw    nJuqx    uPkpA    ZZLWA    AMonU    cyHKs    GuzQO    ZkQxz   
2    ffsLD    nJuqx    uPkpA    qmAUD    AMonU    cyHKs    GuzQO    ZkQxz   
3    ffsLD    nJuqx    uPkpA    ZZLWA    oKdrT    cyHKs    GuzQO    ZkQxz   
4    ffsLD    nJuqx    uPkpA    qmAUD    AMonU    cyHKs    GELir    ZkQxz   

  aCfsveTu EfkPrfXa NrvxpdMQ  nEsgxvAq FcekeISI NmAVTtfA YTdCRVJt QyBloWXZ  \
0    kgmhs    vdDgH    sdZwV       -16    jkpcP    buans    SDrXD    LKzoc   
1    kgmhs    vdDgH    sdZwV       -10    AsSEW    buans    SDrXD    LKzoc   
2    kgmhs    vdDgH    sdZwV        -4    AsSEW    buans    dROpA    LKzoc   
3    kgmhs    vdDgH    sdZwV        -4    jkpcP    buans    dROpA    LKzoc   
4    kgmhs    vdDgH    sdZwV        -4    AsSEW    buans    dROpA    LKzoc   

  NGOnRdqc HKMQJANN ZRrposmO wakWLjkG vmZttwFZ dkoIJCbY NrUWfvEq WqhniYIc  \
0    aDHDh    nqAId    ywnPq    DBYFB    hACfi    fdbOn    hWZvm    IbIUY   
1    aDHDh    ObBVP    ywnPq    AtJzc    hACfi    fdbOn    hWZvm    IbIUY   
2    aDHDh    ObBVP    ywnPq    DBYFB    hACfi    fdbOn    BnEzq    IbIUY   
3    aDHDh    nqAId    ZMPHr    DBYFB    hACfi    fdbOn    hWZvm    IbIUY   
4    aDHDh    ObBVP    ZMPHr    DBYFB    hACfi    fdbOn    BnEzq    IbIUY   

  yHbEDILT EJgrQqET nGMEgWyl IIEHQNUc HfKRIwMb NRVuZwXK UCAmikjV UGbBCHRE  \
0    vfweT    wdCvs    SokyU    uVrbm    QFawF    gLHsu    ycFDT    UbtLs   
1    vfweT    KdnLw    SokyU    uVrbm    UdKcj    gLHsu    ycFDT    UbtLs   
2    vfweT    KdnLw    SokyU    DXLhU    QFawF    gLHsu    oRtiU    UbtLs   
3    vfweT    wdCvs    SokyU    uVrbm    QFawF    gLHsu    oRtiU    UbtLs   
4    vfweT    wdCvs    SokyU    uVrbm    QFawF    gLHsu    oRtiU    UbtLs   

  uJYGhXqG bxKGlBYX nCzVgxgY ltcNxFzI WbEDLWBH IBPMYJlv MxOgekdE ggNglVqE  \
0    WQPIf    FWMHx    DkINd    xoMGF    QKdak    ZUFtA    Cttil    vJxeZ   
1    GJSmz    FWMHx    DkINd    aaPgB    QKdak    ZUFtA    Cttil    nmBHL   
2    GJSmz    FWMHx    DkINd    aaPgB    QKdak    ZUFtA    Cttil    vJxeZ   
3    GJSmz    bDOph    DkINd    xoMGF    QKdak    ZUFtA    Cttil    nmBHL   
4    GJSmz    FWMHx    DkINd    xoMGF    QKdak    ZUFtA    Cttil    nmBHL   

  YDgWYWcJ SqGRfEuW WiwmbjGW benRXROb cOSBrarW JwtIxvKg lRGpWehf dSALvhyd  \
0    ESZVQ    QiWxT    iHtOk    jzcPS    vLHhC    WnTLO    mwImW    mbAAs   
1    ESZVQ    jGNyr    iHtOk    jzcPS    vLHhC    WnTLO    mwImW    mbAAs   
2    ESZVQ    KFwaq    MLQXR    jzcPS    vLHhC    kaiUF    mwImW    mbAAs   


In [4]:
poverty_test.head()

id wBXbHZmp SlDKnCuu KAJOWiiw DsKacCdL rtPrBBPl tMJrvvut jdetlNNF  \
0  10975    JhtDR    GUusz    TuovO    ZYabk    feupP    PHMVg    NDTCU   
1   9346    JhtDR    GUusz    TuovO    ZYabk    feupP    PHMVg    NDTCU   
2  23899    JhtDR    alLXR    BIZns    ZYabk    feupP    PHMVg    NDTCU   
3  26452    JhtDR    GUusz    BIZns    ZYabk    feupP    PHMVg    NDTCU   
4  67507    JhtDR    alLXR    TuovO    ZYabk    feupP    PHMVg    NDTCU   

  maLAYXwi vwpsXRGk sArDRIyX goxNwvnG TYhoEiNm bgfNZfcj sYIButva VZtBaoXL  \
0    sehIp    sPNOc    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   
1    cLAGr    RpXJc    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   
2    sehIp    yygvO    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    usVYt   
3    cLAGr    RpXJc    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    usVYt   
4    TcuXF    ZCIYy    MwLvg    wSMgG    sdlSU    aSGQn    nMLtq    AtADK   

  GUvFHPNA zFkComtB fxbqfEWb HDMHzGif zzwlWZZC nGTepfos DxLvCGgv CbABToOI  \
0    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    GEOXc    tIJga   
1    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    VbCFp    oHRvd   
2    JGQEQ    JETTQ    WobIQ    PSbhz    CtanB    nvuhC    VbCFp    tIJga   
3    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    GEOXc    oHRvd   
4    JGQEQ    zlEqv    WobIQ    PSbhz    CtanB    nvuhC    GEOXc    tIJga   

  qgMygRvX fhKiXuMY uSKnVaKV hESBInAl nzTeWUeM BbKZUYsB UCnazcxd hTraVEWP  \
0    Bdpcw    nJuqx    uPkpA    ZZLWA    AMonU    cyHKs    GuzQO    ZkQxz   
1    ffsLD    nJuqx    uPkpA    ZZLWA    oKdrT    cyHKs    GuzQO    ZkQxz   
2    ffsLD    nJuqx    uPkpA    qmAUD    AMonU    cyHKs    GuzQO    ZkQxz   
3    vswDo    nJuqx    HNHAn    qmAUD    oKdrT    cyHKs    GuzQO    ZkQxz   
4    ffsLD    nJuqx    uPkpA    ZZLWA    AMonU    cyHKs    GuzQO    ZkQxz   

  aCfsveTu EfkPrfXa NrvxpdMQ  nEsgxvAq FcekeISI NmAVTtfA YTdCRVJt QyBloWXZ  \
0    kgmhs    vdDgH    sdZwV        -4    AsSEW    buans    dROpA    LKzoc   
1    kgmhs    vdDgH    sdZwV        -4    jkpcP    buans    dROpA    LKzoc   
2    kgmhs    vdDgH    sdZwV       -16    AsSEW    srfRJ    SDrXD    LKzoc   
3    kgmhs    vdDgH    sdZwV       -16    jkpcP    buans    dROpA    LKzoc   
4    kgmhs    vdDgH    sdZwV        -4    AsSEW    buans    dROpA    WUJmr   

  NGOnRdqc HKMQJANN ZRrposmO wakWLjkG vmZttwFZ dkoIJCbY NrUWfvEq WqhniYIc  \
0    aDHDh    ObBVP    ZMPHr    AtJzc    hACfi    fdbOn    hWZvm    IbIUY   
1    aDHDh    nqAId    ywnPq    DBYFB    hACfi    fdbOn    hWZvm    IbIUY   
2    aDHDh    ObBVP    ywnPq    AtJzc    hACfi    fdbOn    BnEzq    IbIUY   
3    aDHDh    nqAId    ZMPHr    DBYFB    hACfi    fdbOn    BnEzq    IbIUY   
4    aDHDh    ObBVP    ywnPq    DBYFB    hACfi    fdbOn    hWZvm    IbIUY   

  yHbEDILT EJgrQqET nGMEgWyl IIEHQNUc HfKRIwMb NRVuZwXK UCAmikjV UGbBCHRE  \
0    vfweT    wdCvs    SokyU    uVrbm    QFawF    gLHsu    oRtiU    UbtLs   
1    vfweT    wdCvs    SokyU    uVrbm    QFawF    gLHsu    oRtiU    UbtLs   
2    vfweT    wdCvs    SokyU    uVrbm    QFawF    sbeOr    ycFDT    UbtLs   
3    vfweT    wdCvs    SokyU    uVrbm    QFawF    Lgwgt    oRtiU    UbtLs   
4    vfweT    wdCvs    SokyU    uVrbm    QFawF    sbeOr    ycFDT    UbtLs   

  uJYGhXqG bxKGlBYX nCzVgxgY ltcNxFzI WbEDLWBH IBPMYJlv MxOgekdE ggNglVqE  \
0    GJSmz    FWMHx    DkINd    eGKYm    QKdak    ZUFtA    Cttil    nmBHL   
1    GJSmz    FWMHx    DkINd    aaPgB    QKdak    ZUFtA    Cttil    nmBHL   
2    GJSmz    FWMHx    DkINd    eGKYm    QKdak    ZUFtA    Cttil    nmBHL   
3    GJSmz    FWMHx    DkINd    aaPgB    QKdak    ZUFtA    Cttil    nmBHL   
4    GJSmz    FWMHx    DkINd    eGKYm    QKdak    ZUFtA    Cttil    nmBHL   

  YDgWYWcJ SqGRfEuW WiwmbjGW benRXROb cOSBrarW JwtIxvKg lRGpWehf dSALvhyd  \
0    ESZVQ    QiWxT    iHtOk    jzcPS    vLHhC    kaiUF    mwImW    mbAAs   
1    ESZVQ    jGNyr    iHtOk    jzcPS    vLHhC    WnTLO    mwImW    mbAAs   
2    ESZVQ    jGNyr    MLQXR    jzcPS    vLHhC    kaiUF    mwImW    mbAAs   


In [5]:
poverty_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6578 entries, 0 to 6577
Columns: 346 entries, id to country
dtypes: bool(1), int64(5), object(340)
memory usage: 17.3+ MB


In [6]:
poverty_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1625 entries, 0 to 1624
Columns: 346 entries, id to country
dtypes: bool(1), int64(5), object(340)
memory usage: 4.3+ MB


From looking at the info, there are 3 types of features in the dataset, 1 bool, 5 numerical, and 340 categorical.  I'm going to assume that the poverty label is the boolean datatype, but let's check.

In [7]:
poverty_train.select_dtypes(include=['bool']).head()

poor
0  False
1  False
2   True
3   True
4  False

Now that the target label has been identified, let's check for duplicates and nulls.

In [8]:
poverty_train.shape

(6578, 346)

In [9]:
poverty_train.drop_duplicates()
poverty_train.shape

(6578, 346)

In [10]:
poverty_train.dropna(inplace=True)
poverty_train.shape

(6578, 346)

It doesn't look like there are any null values or duplicates in the training dataset.  Let's check the correlation of the numerical features with the label.

In [11]:
def printCorrDesc(data, label):
    """
    Prints the correlation of the features to the label in descending order.
    """
    corr_matrix = data.corr()
    return corr_matrix[label].sort_values(ascending=False)

In [12]:
printCorrDesc(poverty_train, 'poor')

poor        1.000000
nEsgxvAq    0.329241
id         -0.001770
OMtioXZZ   -0.142636
YFMZwKrU   -0.269284
TiwRslOh   -0.291140
Name: poor, dtype: float64

Let's separate the categorical and numerical features now.

In [13]:
#cat_features = list(poverty_train.select_dtypes(include=['object']).drop(['country'], axis=1).columns)
#num_features = list(poverty_train.select_dtypes(include=['int64']).drop(['id'], axis=1).columns)

In [14]:
"""
cat_count_dict = {}
multi_cat_feats = []
ordinal_cat_feats = []
for column in cat_features:
    col = getattr(poverty_train, column)
    cat_count_dict[column] = len(col.value_counts())

for k,v in cat_count_dict.items():
    if v > 10:
        multi_cat_feats.append(k)
    else:
        ordinal_cat_feats.append(k)
"""

'\ncat_count_dict = {}\nmulti_cat_feats = []\nordinal_cat_feats = []\nfor column in cat_features:\n    col = getattr(poverty_train, column)\n    cat_count_dict[column] = len(col.value_counts())\n\nfor k,v in cat_count_dict.items():\n    if v > 10:\n        multi_cat_feats.append(k)\n    else:\n        ordinal_cat_feats.append(k)\n'

Note: My initial thought here was to compile a list of all the categorical features with value counts of more than 5 or 10. I would then treat lower dimensional categorical features as ordinal and the higher ones as nominal.  However, I didn't see a significant improvement (or any improvement at all) in log_loss in doing this. I am noting this here for future exploration.

# Data Preprocessing

In [38]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import log_loss
from sklearn.preprocessing import StandardScaler, LabelBinarizer, LabelEncoder
from sklearn.svm import SVC
import xgboost as xgb
from xgboost.sklearn import XGBClassifier
#from sklearn.pipeline import Pipeline
#from sklearn_pandas import DataFrameMapper

In [51]:
def preprocess_data(data):
    print('Shape before: {}'.format(data.shape))
    data = data.drop('id', axis=1)
    for col in data.columns:
        if data.dtypes[col] == 'object':
            le = LabelBinarizer()
            le.fit(data[col])
            data[col] = le.transform(data[col])
    
        #elif data.dtypes[col] == 'int64':
         #   scaler = StandardScaler()
          #  temp = data[col].values.reshape(-1, 1)
           # scaler.fit(temp)
            #data[col] = scaler.transform(temp)
    print('Shape after: {}:'.format(data.shape))
    return data

I initially tried standardizing all the numerical categories, but I was actually getting worse results when I was doing so. 

In [52]:
poverty_train_transformed = preprocess_data(poverty_train)

Shape before: (6578, 346)
Shape after: (6578, 345):


In [53]:
poverty_test_transformed = preprocess_data(poverty_test)

Shape before: (1625, 346)
Shape after: (1625, 345):


In [54]:
# Change the labels to 0 and 1
poverty_train_transformed['poor'] = poverty_train_transformed.poor.astype('int')
poverty_test_transformed['poor'] = poverty_test_transformed.poor.astype('int')

In [55]:
X_train = poverty_train_transformed.drop(['poor', 'country'], axis=1)
X_test = poverty_test_transformed.drop(['poor', 'country'], axis=1)
y_train = poverty_train_transformed['poor']
y_test = poverty_test_transformed['poor']

# Initial Modeling and Getting Feature Importances

My initial thought is to run a logistic regression and see what my log loss is.

In [56]:
log_r = LogisticRegression(penalty='l1')
log_r.fit(X_train, y_train)
y_pred = log_r.predict_proba(X_test)
log_loss(y_test, y_pred)

0.30783516607142708

Now that I have an initial baseline metric I want to improve, I use Random Forest to grab the important features and decide which ones I want to remove.

In [57]:
rfc = RandomForestClassifier(n_estimators=50, max_depth=10)
rfc.fit(X_train, y_train)
y_pred = rfc.predict_proba(X_test)
log_loss(y_test, y_pred)

0.43258968924762303

In [58]:
feature_importances = pd.DataFrame(rfc.feature_importances_,
                                   index = X_train.columns,
                                   columns=['importance']).sort_values('importance', ascending=False)

unimportant_features = feature_importances[feature_importances['importance'] < 0.000001].index

In [59]:
poverty_train_transformed.drop(unimportant_features, axis=1, inplace=True)
poverty_test_transformed.drop(unimportant_features, axis=1, inplace=True)

In [60]:
poverty_train_transformed.shape

(6578, 325)

After determining some of the unimportant features, I have removed about 20 features.  Now let's check to see if the models performed better.

# Check for Improvements

In [61]:
X_train = poverty_train_transformed.drop(['poor', 'country'], axis=1)
X_test = poverty_test_transformed.drop(['poor', 'country'], axis=1)
y_train = poverty_train_transformed['poor']
y_test = poverty_test_transformed['poor']

log_r = LogisticRegression(penalty='l1', solver='liblinear')
log_r.fit(X_train, y_train)
y_pred = log_r.predict_proba(X_test)
log_loss(y_test, y_pred)

0.27290012422493576

I can see that my Logistic Regression log loss improved from about ~0.31 to ~0.27.

In [62]:
svc = SVC(C=1, probability=True, verbose=True)
svc.fit(X_train, y_train)
y_pred = svc.predict_proba(X_test)
log_loss(y_test, y_pred)

[LibSVM]

0.27017207353263595

Using a Support Vector Classifier model gives me about the same result.  

# Gradient Boosting with XGBoost

It should be noted at the current time, I do not fully understand extreme gradient boosting and I am simply fitting a model for improvement purposes.

In [63]:
xgb1 = XGBClassifier(learning_rate =0.03,
                     n_estimators=1500,
                     max_depth=3,
                     min_child_weight=1,
                     gamma=0.3, 
                     silent=False,
                     subsample=0.7,
                     objective= 'binary:logistic',
                     nthread=4,
                     scale_pos_weight=1,
                     reg_alpha=0.05)

In [64]:
xgb1.fit(X_train, y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0.3, learning_rate=0.03, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=1500,
       n_jobs=1, nthread=4, objective='binary:logistic', random_state=0,
       reg_alpha=0.05, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=False, subsample=0.7)

In [65]:
y_pred = xgb1.predict_proba(X_test)

In [66]:
log_loss(y_test, y_pred)

0.25744970417425672

Testing some of the parameters has been giving log losses ranging from 0.25 ~ 0.26, which is a slight improvement over the other models.  I could run a GridSearch, but in the interest of time, I will leave that for future options.

# Future ideas / things to improve on

* More feature engineering, trying different ratios or new extractions of the features.
* Recursive feature elimination.
* GridSearch or RandomSearch for parameter fine-tuning (I didn't do this in the interest of time).
* Create some pipelines to clean up my code.
* Find more effective ways to scale or standardize my numerical features.
* Explore other methods of Classification
* Possibly cluster the existing data into groups with unsupervised learning before testing supervised.
* Learn NN's